unittests/test-contracts/savanna/ibc/ibc.cpp

```diff
@@ -84,98 +77,77 @@ void svnn_ibc::_maybe_remove_from_cache(){
 
     //fetch the finalizer policy where generation num is equal prepare.input.finalizer_policy_generation, and that it is still in force
     auto itr = _policies_table.find(finalizer_policy_generation);
     check(itr!=_policies_table.end(), "finalizer policy not found");
-    storedpolicy target_policy = *itr;
-    auto fa_itr = target_policy.finalizers.begin();
-    auto fa_end_itr = target_policy.finalizers.end();
-    size_t finalizer_count = std::distance(fa_itr, fa_end_itr);
-    bitset b(finalizer_count, qc.finalizers);
-
-    bool first = true;
-
-    size_t index = 0;
-    uint64_t weight = 0;
-
-    bls_g1 agg_pub_key;
-
-    while (fa_itr != fa_end_itr){
-        if (b.test(index)){
-            bls_g1 pub_key = decode_bls_public_key_to_g1(fa_itr->public_key);
-            if (first){
-                first=false;
-                agg_pub_key = pub_key;
-            }
-            else agg_pub_key = _g1add(agg_pub_key, pub_key);
-            weight+=fa_itr->weight;
-        }
-        index++;
-        fa_itr++;
-    }
+    return *itr;
+
+}
+
+void ibc::_check_finality_proof(const finality_proof& finality_proof, const block_proof_of_inclusion& target_block_proof_of_inclusion){
+
+    //attempt to retrieve the stored policy with the correct generation number
+    finalizer_policy_input finalizer_policy = _get_stored_finalizer_policy(finality_proof.qc_block.finalizer_policy_generation);
+
+    //verify QC. If QC is valid, it means that we have reaced finality on the block referenced by the finality_mroot
+    _check_qc(finality_proof.qc, block_finality_data_internal(finality_proof.qc_block).finality_digest(), finalizer_policy);
+
+    //check if the target proof of inclusion correctly resolves to the root of the finality proof
+    _check_target_block_proof_of_inclusion(target_block_proof_of_inclusion, finality_proof.qc_block.finality_mroot);
+    
+    //if the finality_mroot we just proven is more recent than the last root we have stored, store it
+    uint64_t offset = target_block_proof_of_inclusion.final_block_index - target_block_proof_of_inclusion.target_block_index;
+
+    dynamic_data_v0 d_data;
 
-    //verify that we have enough vote weight to meet the quorum threshold of the target policy
-    check(weight>=target_policy.threshold, "insufficient signatures to reach quorum");
-    std::array<uint8_t, 32> fd_data = finality_digest.extract_as_byte_array();
-    std::string message(fd_data.begin(), fd_data.end());
+    if (std::holds_alternative<extended_block_data>(target_block_proof_of_inclusion.target)) d_data = std::get<extended_block_data>(target_block_proof_of_inclusion.target).dynamic_data;
+    else if (std::holds_alternative<simple_block_data>(target_block_proof_of_inclusion.target)) d_data = std::get<simple_block_data>(target_block_proof_of_inclusion.target).dynamic_data;
+    else check(false, "invalid block data");

```

Comment:
I think this can be replaced with:

`dynamic_data_v0 d_data = std::visit([&](const auto& bd) { return bd.dynamic_data; }, target_block_proof_of_inclusion.target);`